### DISCUSSION TITLE
Using Dashboard with Heroku

### DISCUSSION THREAD

**Discussion by marcphilippebeaujean-abertay** [jobrunr/jobrunr#249] [Q&A] [answered]
Since the dashboard is running on a different than port than my application (spring for example), is there any way to access the dashboard, like using it as a self-hosted application? I would love to have visibility into how my Jobs are performing! 
PS: Thanks Ronald for a great tool.

**Comment by rdehuyss** [comment, marked as answer]
Well, Heroku has the constraint that you can not do any port configuration, that's a pity...

What you could do is embed a simple reverse proxy in your spring app. See https://stackoverflow.com/questions/29359254/simple-reverse-proxy-with-spring-boot-and-netflix-zuul on how to do so.

**Reply by benjamin-thomas** [reply]
Hello,

Congrats on your library, I appreciate the nice and simple design :)

I'm having the same problem, working on a spring boot toy project and deploying to Heroku.

Unfortunately, this solution seems to be a no-go in 2022 as Zuul isn't supported by the framework anymore.

https://stackoverflow.com/questions/72952820/does-zuul-have-any-support-issues-in-springboot-2-7-and-java-11

I tried to implement my own proxying but this doesn't look like a simple job, considering the GUI is a SPA with it's own idea of what routing should be :(

I'm not sure what could be a good solution, because I feel that Heroku's restrictions are actually a good thing and I'm not so fond of the duplicated authentication logic.

If the library could somehow return a basic dashboard as a dump of raw html, that would probably make embedding the dashboard into one's app much easier. At least for a read only view which would be better than nothing.

What do you think @rdehuyss?

**Reply by rdehuyss** [reply]
Hhmmm, I don't know yet how I feel about that to be honest. 

Do you feel like contributing something like that (we'll need to discuss this though before you put time in)?

The simplest thing you could do is run a JobRunr Dashboard instance (only 1) and expose the dashboard on port 8080.

**Reply by benjamin-thomas** [reply]
> The simplest thing you could do is... expose the dashboard on port 8080

I'm not sure I got you here. I can't use port 8080 in my dev env since my spring boot dev server already uses this port. And I can't use port 8080 in prod since Heroku won't allow it. Internally, a random port is chosen by the infrastructure, so my app's server has to bind to a random port, exposed by a `PORT` env var. That's because my app is then already proxied by another server that listens to port 443, and handles certificates, etc. So, to work well and benefit from this kind of environment, one's app shouldn't depend on custom ports configuration, etc.

I can't make promises about contributing, but I could at least put in a few hours and see how far I go. I was merely curious to hear if I was misunderstanding something basic.

If your're ok revisiting the decision to have a separate server, I'd favor tackling at the bigger picture (though I've seen elsewhere your motivation for doing this is to reduce the maintenance burden of keeping things compatible over time)

I'm trying to get into Java and its ecosystem so I'll probably have a few questions (I'm coming from a rails background).

I've cloned your project to have a look, and I'm not sure how to proceed to setup and get a dev workflow going so that'll be my first question :)

**Reply by rdehuyss** [reply]
What I mean is to add a really small node in your heroku setup which only runs the dashboard (disable your spring server, disable jobrunr background job processing). Use the $PORT variable from Heroku to pass it to JobRunr Dashboard and all should be well :-)

**Reply by benjamin-thomas** [reply]
Thanks for the suggestion.

Disabling the web server doesn't look like a trivial task unfortunately, since a lot of code tries to load regardless. Which I would have to either suppress, or delete at which point I'm back at square one in terms of complexity (as a library consumer).

I think I'll query the db and implement a few simple endpoints myself. It'll probably be less work and less brittle for me going forward :)
